### PR TITLE
[SPARK-15454][SQL] Filter out files starting with _

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

Many other systems (e.g. Impala) uses _xxx as staging, and Spark should not be reading those files.
## How was this patch tested?

Added a unit test case.

### CODE REVIEW

**Comment by rxin** [comment]
cc @liancheng

**Comment by marmbrus** [comment]
LGTM, pending tests.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala [hunk 1]
```diff
@@ -341,11 +341,11 @@ private[sql] object HadoopFsRelation extends Logging {
...
-    // TODO: We should try to filter out all files/dirs starting with "." or "_".
-    // The only reason that we are not doing it now is that Parquet needs to find those
-    // metadata files from leaf files returned by this methods. We should refactor
-    // this logic to not mix metadata files with data files.
-    pathName == "_SUCCESS" || pathName == "_temporary" || pathName.startsWith(".")
+    // We filter everything that starts with _ and ., except _common_metadata and _metadata
+    // because Parquet needs to find those metadata files from leaf files returned by this method.
+    // We should refactor this logic to not mix metadata files with data files.
+    (pathName.startsWith("_") || pathName.startsWith(".")) &&
+      !pathName.startsWith("_common_metadata") && !pathName.startsWith("_metadata")
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: Why `startsWith` instead of `==` here?

**Comment by liancheng** [comment]
LGTM except for one minor comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: just in case we do other variants here ..

**Comment by rxin** [comment]
Merging in master/2.0. Thanks.

**Pull request closed without merging**